File: framework/languages/dotnet/samples/TicTacToe/Server/Play/Domain/TicTacToe/TicTacToeMatch.cs
using TicTacToe.Shared.Contracts;

namespace TicTacToe.Server.Play.Domain.TicTacToe;

internal sealed class TicTacToeMatch(string roomId, TimeSpan turnTimeout)
{
    private const string MarkX = "X";
    private const string MarkO = "O";
    private const string WaitingForPlayers = "WaitingForPlayers";
    private const string InProgress = "InProgress";
    private const string Won = "Won";
    private const string Draw = "Draw";
    private const string TurnTimedOut = "TurnTimedOut";

    private readonly Dictionary<string, string> _players = new(StringComparer.Ordinal);
    private readonly TicTacToeBoard _board = new();
    private string _nextTurn = MarkX;
    private string _status = WaitingForPlayers;
    private string? _winner;
    private string? _lastMoveActorId;
    private int? _lastMoveCell;
    private DateTimeOffset? _turnDeadline;

    public TicTacToeJoinChange JoinPlayer(string actorId, DateTimeOffset now)
    {
        if (_players.TryGetValue(actorId, out var existingMark))
        {
            return new TicTacToeJoinChange(Snapshot(), existingMark, false);
        }

        var mark = _players.Count switch
        {
            0 => MarkX,
            1 => MarkO,
            _ => throw new InvalidOperationException("Tic-tac-toe game already has two players.")
        };

        _players.Add(actorId, mark);
        if (_status == WaitingForPlayers && _players.Count == 2)
        {
            _status = InProgress;
            ResetTurnDeadline(now);
        }

        return new TicTacToeJoinChange(Snapshot(), mark, true);
    }

    public TicTacToeMoveChange PlaceMark(string actorId, int cell, DateTimeOffset now)
    {
        if (!_players.TryGetValue(actorId, out var mark))
        {
            throw new InvalidOperationException("Player has not joined this game.");
        }

        if (_status != InProgress)
        {
            throw new InvalidOperationException($"Game is not in progress. status={_status}");
        }

        if (!string.Equals(mark, _nextTurn, StringComparison.Ordinal))
        {
            throw new InvalidOperationException($"It is {_nextTurn}'s turn.");
        }

        _board.Place(mark, cell);
        _lastMoveActorId = actorId;
        _lastMoveCell = cell;
        AdvanceAfterMove(actorId, mark, now);
        return new TicTacToeMoveChange(Snapshot());
    }

    public TicTacToeTickChange Tick(DateTimeOffset now)
    {
        if (_status != InProgress
            || _turnDeadline is not { } deadline
            || now < deadline)
        {
            return new TicTacToeTickChange(Snapshot(), false);
        }

        var timedOut = _players.FirstOrDefault(player => player.Value == _nextTurn).Key;
        var winner = _players.FirstOrDefault(player => player.Value != _nextTurn).Key;

        _status = TurnTimedOut;
        _winner = string.IsNullOrEmpty(winner) ? null : winner;
        _nextTurn = string.Empty;
        _lastMoveActorId = string.IsNullOrEmpty(timedOut) ? null : timedOut;
        _lastMoveCell = null;
        _turnDeadline = null;
        return new TicTacToeTickChange(Snapshot(), true);
    }

    public GameState Snapshot()
    {
        var x = _players.FirstOrDefault(static player => player.Value == MarkX).Key;
        var o = _players.FirstOrDefault(static player => player.Value == MarkO).Key;
        return new GameState(
            roomId,
            _board.Snapshot(),
            _status,
            _winner,
            _nextTurn,
            string.IsNullOrEmpty(x) ? null : x,
            string.IsNullOrEmpty(o) ? null : o,
            _lastMoveActorId,
            _lastMoveCell);
    }

    private void AdvanceAfterMove(string actorId, string mark, DateTimeOffset now)
    {
        if (_board.HasWon(mark))
        {
            _status = Won;
            _winner = actorId;
            _nextTurn = string.Empty;
            _turnDeadline = null;
            return;
        }

        if (_board.IsFull)
        {
            _status = Draw;
            _winner = null;
            _nextTurn = string.Empty;
            _turnDeadline = null;
            return;
        }

        _nextTurn = mark == MarkX ? MarkO : MarkX;
        ResetTurnDeadline(now);
    }

    private void ResetTurnDeadline(DateTimeOffset now)
    {
        _turnDeadline = now.Add(turnTimeout);
    }

}

internal sealed record TicTacToeJoinChange(
    GameState State,
    string Mark,
    bool IsNewPlayer);

internal sealed record TicTacToeMoveChange(GameState State);

internal sealed record TicTacToeTickChange(
    GameState State,
    bool HasChanged);

File: framework/languages/node/samples/TicTacToe.Ts/Server/Play/Domain/TicTacToe/tictactoe-match.ts
const { TicTacToeBoard } = require('./tictactoe-board');
import type { GameState, TicTacToeActor } from '../../../../Shared/Contracts/messages';

type JoinedPlayer = {
  actorId: string;
  mark: string;
  actor: TicTacToeActor;
};

type TicTacToeJoinChange = {
  state: GameState;
  joined: JoinedPlayer;
  newlyJoined: boolean;
};

type TicTacToeMoveChange = {
  state: GameState;
};

type TicTacToeTickChange = {
  state: GameState;
  changed: boolean;
};

class TicTacToeMatch {
  [key: string]: any;
  constructor(roomId: string, turnTimeoutMs: number = 15000) {
    this.roomId = roomId;
    this.turnTimeoutMs = turnTimeoutMs;
    this.players = new Map();
    this.board = new TicTacToeBoard();
    this.status = 'WaitingForPlayers';
    this.winner = null;
    this.nextTurn = null;
    this.lastMoveActorId = null;
    this.lastMoveCell = null;
    this.turnDeadline = null;
  }

  joinPlayer(actor: TicTacToeActor): TicTacToeJoinChange {
    const existing = this.players.get(actor.actorId);
    if (existing !== undefined) {
      return { state: this.snapshot(), joined: existing, newlyJoined: false };
    }
    if (this.players.size >= 2) {
      throw new Error(`Room '${this.roomId}' is full.`);
    }
    const mark = this.players.size === 0 ? 'X' : 'O';
    const joined = { actorId: actor.actorId, mark, actor };
    this.players.set(actor.actorId, joined);
    if (this.players.size === 2) {
      this.status = 'InProgress';
      this.nextTurn = this.xActorId();
      this.resetTurnDeadline();
    }
    return { state: this.snapshot(), joined, newlyJoined: true };
  }

  placeMark(actorId: string, cell: number): TicTacToeMoveChange {
    const player = this.players.get(actorId);
    if (player === undefined) {
      throw new Error(`Actor '${actorId}' has not joined room '${this.roomId}'.`);
    }
    if (this.status !== 'InProgress') {
      throw new Error(`Room '${this.roomId}' is not in progress.`);
    }
    if (this.nextTurn !== actorId) {
      throw new Error(`Actor '${actorId}' cannot move out of turn.`);
    }
    this.board.place(player.mark, cell);
    this.lastMoveActorId = actorId;
    this.lastMoveCell = cell;
    this.advanceAfterMove(actorId, player.mark);
    return { state: this.snapshot() };
  }

  tick(now: number = Date.now()): TicTacToeTickChange {
    if (this.status !== 'InProgress' || this.turnDeadline === null || now < this.turnDeadline) {
      return { state: this.snapshot(), changed: false };
    }
    const timedOut = this.nextTurn;
    const winner = [...this.players.values()]
      .find((player) => player.actorId !== timedOut)?.actorId ?? null;
    this.status = 'TurnTimedOut';
    this.winner = winner;
    this.nextTurn = null;
    this.lastMoveActorId = timedOut;
    this.lastMoveCell = null;
    this.turnDeadline = null;
    return { state: this.snapshot(), changed: true };
  }

  snapshot(): GameState {
    return {
      roomId: this.roomId,
      board: this.board.snapshot(),
      status: this.status,
      winner: this.winner,
      nextTurn: this.nextTurn,
      xActorId: this.xActorId(),
      oActorId: this.oActorId(),
      lastMoveActorId: this.lastMoveActorId,
      lastMoveCell: this.lastMoveCell
    };
  }

  xActorId(): string | null {
    return [...this.players.values()].find((player) => player.mark === 'X')?.actorId ?? null;
  }

  oActorId(): string | null {
    return [...this.players.values()].find((player) => player.mark === 'O')?.actorId ?? null;
  }

  advanceAfterMove(actorId: string, mark: string): void {
    if (this.board.hasWon(mark)) {
      this.status = 'Won';
      this.winner = actorId;
      this.nextTurn = null;
      this.turnDeadline = null;
      return;
    }
    if (this.board.isFull()) {
      this.status = 'Draw';
      this.winner = null;
      this.nextTurn = null;
      this.turnDeadline = null;
      return;
    }
    this.nextTurn = [...this.players.values()].find((player) => player.actorId !== actorId).actorId;
    this.resetTurnDeadline();
  }

  resetTurnDeadline(): void {
    this.turnDeadline = Date.now() + this.turnTimeoutMs;
  }
}

export { TicTacToeMatch };
export type { JoinedPlayer, TicTacToeJoinChange, TicTacToeMoveChange, TicTacToeTickChange };
